components/webext-storage/src/sync/incoming.rs

```diff
@@ -0,0 +1,470 @@
+                    IncomingAction::TakeRemote { data }
+                }
+                (None, None) => {
+                    // Nothing anywhere - odd, but OK.
+                    IncomingAction::Same
+                }
+            }
+        }
+        IncomingState::NotLocal { incoming, .. } => {
+            // No local data but there's mirror and an incoming record.
+            // This means a local deletion is being replaced by, or just re-doing
+            // the incoming record.
+            match incoming {
+                Some(data) => IncomingAction::TakeRemote { data },
+                None => IncomingAction::Same,
+            }
+        }
+        IncomingState::IncomingOnly { incoming } => {
+            // Only the staging record exists - this means it's the first time
+            // we've ever seen it. No conflict possible, just take the remote.
+            match incoming {
+                Some(data) => IncomingAction::TakeRemote { data },
+                None => IncomingAction::DeleteLocally,
+            }
+        }
+    }
+}
+
+// Apply the actions necessary to fully process the incoming items.
+pub fn apply_actions<S: ?Sized + Interruptee>(
+    tx: &Transaction<'_>,
+    actions: Vec<(IncomingItem, IncomingAction)>,
+    signal: &S,
+) -> Result<()> {
+    for (item, action) in actions {
+        signal.err_if_interrupted()?;
+
+        log::trace!("action for '{}': {:?}", item.ext_id, action);
+        // XXX - change counter should be updated consistently here!
+        match action {
+            IncomingAction::DeleteLocally => {
+                // Can just nuke it entirely.
+                tx.execute_named_cached(
+                    "DELETE FROM storage_sync_data WHERE ext_id = :ext_id",
+                    &[(":ext_id", &item.ext_id)],
+                )?;
+            }
+            // We want to update the local record with 'data' and after this update the item no longer is considered dirty.
+            IncomingAction::TakeRemote { data } => {
+                tx.execute_named_cached(
+                    "UPDATE storage_sync_data SET data = :data, sync_change_counter = 0 WHERE ext_id = :ext_id",
+                    &[
+                        (":ext_id", &item.ext_id),
+                        (":data", &serde_json::Value::Object(data).as_str()),
+                    ]
+                )?;
+            }
+
+            // We merged this data, so need to update locally but still consider
+            // it dirty because the merged data must be uploaded.
+            IncomingAction::Merge { data } => {
+                tx.execute_named_cached(
+                    "UPDATE storage_sync_data SET data = :data, sync_change_counter = sync_change_counter + 1 WHERE ext_id = :ext_id",
+                    &[
+                        (":ext_id", &item.ext_id),
+                        (":data", &serde_json::Value::Object(data).to_string()),
```

Comment:
Micro-nit: It looks like Rusqlite will [automatically implement `ToSql`](https://github.com/rusqlite/rusqlite/blob/d88fe1c1b18f308b85f38ba0cf89e1dff6998638/src/types/mod.rs#L61-L62) if it's compiled with the `serde_json` feature? Paging @thomcc to make sure! ☎️

Reply:
> Micro-nit: It looks like Rusqlite will [automatically implement `ToSql`](https://github.com/rusqlite/rusqlite/blob/d88fe1c1b18f308b85f38ba0cf89e1dff6998638/src/types/mod.rs#L61-L62) if it's compiled with the `serde_json` feature?

It does indeed - nice catch!